drivers/platform/stm32/stm32_gpio_irq.c

```diff
@@ -0,0 +1,512 @@
+ */
+void HAL_GPIO_EXTI_Callback(uint16_t pin)
+{
+	int ret;
+	struct irq_action *action;
+	struct irq_action key;
+
+	key.irq_id = no_os_find_first_set_bit(pin);
+
+	if (EXTI->PR & pin) {
+		/* Clear pending bit */
+		EXTI->PR = pin;
+
+		ret = no_os_list_read_find(actions, (void **)&action, &key);
+		if (ret)
+			return;
+
+		if (action->callback)
+			action->callback(action->ctx);
+	}
+}
+
+/**
+ * @brief Initialize the GPIO interrupt controller
+ * @param desc - Pointer where the configured instance is stored
+ * @param param - Configuration information for the instance
+ * @return 0 in case of success, errno error codes otherwise.
+ */
+static int32_t stm32_gpio_irq_ctrl_init(struct no_os_irq_ctrl_desc **desc,
+					const struct no_os_irq_init_param *param)
+{
+	static struct no_os_irq_ctrl_desc *gpio_irq_desc_arr[STM32_IRQ_CTRL_NB];
+	static bool initialized[STM32_IRQ_CTRL_NB] =  {false};
+
+	int ret;
+	struct no_os_irq_ctrl_desc *gpio_irq_desc;
+	struct stm32_gpio_irq_desc *sdesc;
+	struct stm32_gpio_irq_init_param *gpio_irq_ip;
+	uint8_t gpio_irq_ctrl_id;
+
+	if (!param || !param->extra)
+		return -EINVAL;
+
+	gpio_irq_ip =  param->extra;
+
+	/* Check port number */
+	if (!IS_EXTI_GPIO_PORT(gpio_irq_ip->port_nb))
+		return -EINVAL;
+
+	/* Check pin number */
+	if (!IS_EXTI_GPIO_PIN(param->irq_ctrl_id))
+		return -EINVAL;
+
+	/* Obtain gpio irq controller id */
+	if (param->irq_ctrl_id <= 4)
+		gpio_irq_ctrl_id = param->irq_ctrl_id;
+	else if (param->irq_ctrl_id >= 5 && param->irq_ctrl_id <= 9)
+		gpio_irq_ctrl_id = 5;
+	else
+		gpio_irq_ctrl_id = 6;
+
+	if (!initialized[gpio_irq_ctrl_id]) {
+
+		gpio_irq_desc = calloc(1, sizeof(*gpio_irq_desc));
+		if (!gpio_irq_desc)
+			return -ENOMEM;
+
+		sdesc = (struct stm32_gpio_irq_desc*)calloc(1,
+				sizeof(struct stm32_gpio_irq_desc));
+		if (!sdesc) {
+			ret = -ENOMEM;
+			goto error;
+		}
+		/* Add port number */
+		sdesc->port_nb = gpio_irq_ip->port_nb;
+
+		gpio_irq_desc->extra = sdesc;
+		gpio_irq_desc->irq_ctrl_id = param->irq_ctrl_id;
+		gpio_irq_desc->platform_ops = param->platform_ops;
+
+		ret = no_os_list_init(&actions, NO_OS_LIST_PRIORITY_LIST, irq_action_cmp);
+		if (ret)
+			goto error;
+
+		gpio_irq_desc_arr[gpio_irq_ctrl_id] = gpio_irq_desc;
+		initialized[gpio_irq_ctrl_id] = true;
+	}
+
+	*desc = gpio_irq_desc_arr[gpio_irq_ctrl_id];
+
+	return 0;
+error:
+	no_os_list_remove(actions);
+	free(gpio_irq_desc);
+	free(sdesc);
+
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by irq_ctrl_init()
+ * @param desc - Interrupt controller descriptor.
+ * @return 0 in case of success, errno error codes otherwise.
+ */
+static int32_t stm32_gpio_irq_ctrl_remove(struct no_os_irq_ctrl_desc *desc)
+{
+	struct no_os_callback_desc *discard;
+
+	if (!desc)
+		return -EINVAL;
+
+	while (0 == no_os_list_get_first(actions, (void **)&discard))
+		free(discard);
+
+	no_os_list_remove(actions);
+	free(desc->extra);
+	free(desc);
+
+	return 0;
+}
```

Comment:
I think this function needs to set an `initialized` entry as false (the same would apply to the remove function from the NVIC controller).

Reply:
You are right, good catch! Thank you!